src/execute.cxx

```diff
@@ -294,9 +351,17 @@ bool ExecuteCommand::Execute(const std::string &filename)
  */
 int ExecuteCommand::Join()
 {
+    // Allow primary command to conclude
+    // ensures stdout and stderr readers
+    // exit only once primary command process
+    // has concluded
+    // The child and std err pipe readers
+    // will not terminate under normal conditions
+    // unless this process concludes and sets the terminate flag
+    int commandError = this->exit_code_future.get();
```

Comment:
I think this should be rephrased:

* I think calling `.get` here calls `ReportExitCode`, and `->terminated` is not set until that is run
* so this *must* be called first, because the other threads will not terminate until the first thread is resolved

this is slightly different from above e.g. 

> unless this process concludes and sets the terminate flag

If you put `->terminated=true` at the end of `ExecuteCommand::Execute`, then I think this would be accurate (and at that point there wouldn't be sensitivity here in the order of `.get` calls, which might be good.